### PR TITLE
Automatically enable incremental syncs for singer sources

### PR DESCRIPTION
**Pull request by sherifnada**
## What
PoC to auto-detect if a singer source supports incremental and propagate the correct cursor fields and supported sync_modes automagically. 

This has been tested with Shopify and Stripe and is working correctly. The **great** news with this PR is that this means we automatically have incremental for any singer source that supports incremental according to the singer spec. The bad news is I'm not so confident any random singer repo will do this correctly. 

see [here](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#metadata) to understand more about how Singer catalogs express support for incremental syncs

## Reading
* `singer_helpers.py`
* everything else is small fixes (catalog or configs) to make the PR work

## TODO 
- [x] Verify if the singer spec allows for singer taps to incrementally sync without specifying valid bookmark keys. **EDIT**: no formal specification. Will assume answer is no for now. 
- [x] Verify this works with legacy style singer catalogs. **EDIT** update: only difference between legacy catalogs (aka properties) and catalog seems to be that the `selected` property goes on the `schema` field in properties, whereas it goes in the `metadata` in catalog. This is based on scouring the singer slack and various github repos, so please correct it if you know better. 
- [x] We currently don't check if sync_mode is incremental before passing on state information to the singer tap. This should ideally change. **EDIT** we'll pass state info if it's there and let the catalog determine sync mode as usual. 

## Open questions/callouts
* Currently if a singer tap supports more than one possible cursor field, we deterministically pick one. Ideally the user can choose. 
* In a future PR: add standard tests for incremental. It's pretty manual to test at the moment and time consuming. 
* In a future PR: write tests for singer helpers #806 because it is getting non-trivial

### CODE REVIEW

**Review by cgardens — COMMENTED** [review]
this PR made my day. love it.

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 1]
```diff
@@ -73,7 +90,22 @@ def singer_catalog_to_airbyte_catalog(singer_catalog: Dict[str, any]) -> Airbyte
...
             name = stream.get("stream")
             schema = stream.get("schema")
-            airbyte_streams += [AirbyteStream(name=name, json_schema=schema)]
+            airbyte_stream = AirbyteStream(name=name, json_schema=schema)
+
+            metadatas = stream.get("metadata")
+            stream_metadata_container = get_stream_level_metadata(metadatas) if metadatas else None
+            stream_metadata = stream_metadata_container.get("metadata")
+            if stream_metadata:
+                # TODO unclear from the singer spec what behavior should be if there are no valid replication keys, but forced-replication-method is INCREMENTAL.
```
Comment (cgardens) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:99]: i think this is the a good choice. i think most taps throw exceptions in this case.

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py [hunk 2]
```diff
@@ -73,7 +90,22 @@ def singer_catalog_to_airbyte_catalog(singer_catalog: Dict[str, any]) -> Airbyte
...
+                #  For now requiring replication keys for a stream to be considered incremental.
+                replication_keys = stream_metadata.get("valid-replication-keys", [])
+                if len(replication_keys) > 0:
+                    airbyte_stream.source_defined_cursor = True
+                    airbyte_stream.supported_sync_modes = [SyncMode.full_refresh, SyncMode.incremental]
+                    # TODO if there are multiple replication keys, allow configuring which one is used. For now we deterministically take the first
```
Comment (cgardens) [inline airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:105]: how common is this? will we likely get through most of our api integrations without having to do this? i like the approach. let's only bother doing it once we encounter a tap that forces us to.

**Review by cgardens — APPROVED** [review]
awesome.

File: airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py (hunk 2 above)
Comment (sherifnada) [reply airbyte-integrations/bases/base-singer/base_singer/singer_helpers.py:105]: haven't found a source that does multiple keys yet

**Pull request merged**